<?xml version="1.0"?>
<robot xmlns:xacro="http://ros.org/wiki/xacro">

  <ros2_control name="XSHardwareInterface" type="system">

    <hardware>

      <!-- If we are using the actual, physical robot -->
      <xacro:if value="${hardware_type == 'actual'}">
        <plugin>interbotix_xs_ros_control/XSHardwareInterface</plugin>
        <param name="loop_hz">10</param>
        <param name="group_name">arm</param>
        <param name="gripper_name">gripper</param>
        <param name="joint_states_topic">joint_states</param>
      </xacro:if>

      <!-- If doing basic simulation of the robot -->
      <xacro:if value="${hardware_type == 'fake'}">
        <plugin>fake_components/GenericSystem</plugin>
      </xacro:if>

      <!-- If simulating the robot in Gazebo Classic -->
      <xacro:if value="${hardware_type == 'gz_classic'}">
        <plugin>gazebo_ros2_control/GazeboSystem</plugin>
      </xacro:if>

      <!-- If simulating the robot in Gazebo Modern -->
      <xacro:if value="${hardware_type == 'ignition'}">
        <plugin>ign_ros2_control/IgnitionSystem</plugin>
      </xacro:if>

    </hardware>

    <joint name="waist">
      <command_interface name="position">
        <param name="min">"${waist_limit_lower}"</param>
        <param name="max">"${waist_limit_upper}"</param>
      </command_interface>
      <state_interface name="position">
        <param name="initial_value">${waist_initial_value}</param>
      </state_interface>
    </joint>

    <joint name="shoulder">
      <command_interface name="position">
        <param name="min">"${shoulder_limit_lower}"</param>
        <param name="max">"${shoulder_limit_upper}"</param>
      </command_interface>
      <state_interface name="position">
        <param name="initial_value">${shoulder_initial_value}</param>
      </state_interface>
    </joint>

    <joint name="elbow">
      <command_interface name="position">
        <param name="min">"${elbow_limit_lower}"</param>
        <param name="max">"${elbow_limit_upper}"</param>
      </command_interface>
      <state_interface name="position">
        <param name="initial_value">${elbow_initial_value}</param>
      </state_interface>
    </joint>

    <xacro:if value="${dof >= 6}">
      <joint name="forearm_roll">
        <command_interface name="position">
          <param name="min">"${forearm_roll_limit_lower}"</param>
          <param name="max">"${forearm_roll_limit_upper}"</param>
        </command_interface>
        <state_interface name="position">
          <param name="initial_value">${forearm_roll_initial_value}</param>
        </state_interface>
      </joint>
    </xacro:if>

    <joint name="wrist_angle">
      <command_interface name="position">
        <param name="min">"${wrist_angle_limit_lower}"</param>
        <param name="max">"${wrist_angle_limit_upper}"</param>
      </command_interface>
      <state_interface name="position">
        <param name="initial_value">${wrist_angle_initial_value}</param>
      </state_interface>
    </joint>

    <xacro:if value="${dof >= 5}">
      <joint name="wrist_rotate">
        <command_interface name="position">
          <param name="min">"${wrist_rotate_limit_lower}"</param>
          <param name="max">${wrist_rotate_limit_upper}</param>
        </command_interface>
        <state_interface name="position"/>
      </joint>
    </xacro:if>

    <xacro:if value="$(arg use_gripper)">
      <xacro:if value="${hardware_type in ('fake', 'gz_classic', 'ignition')}">
        <joint name="gripper">
          <command_interface name="position"/>
          <state_interface name="position"/>
        </joint>
      </xacro:if>

      <xacro:if value="$(arg show_gripper_fingers)">
        <joint name="left_finger">
          <command_interface name="position">
            <param name="min">"${finger_limit_lower}"</param>
            <param name="max">"${-finger_limit_lower}"</param>
          </command_interface>
          <state_interface name="position">
            <param name="initial_value">${finger_initial_value}</param>
          </state_interface>
        </joint>

        <xacro:if value="${hardware_type in ('gz_classic', 'ignition')}">
          <joint name="right_finger">
            <!-- <param name="mimic">left_finger</param>
            <param name="multiplier">1</param> -->
            <command_interface name="position">
              <param name="min">"${-finger_limit_lower}"</param>
              <param name="max">"${finger_limit_lower}"</param>
            </command_interface>
            <state_interface name="position">
              <param name="initial_value">-${finger_initial_value}</param>
            </state_interface>
          </joint>  
        </xacro:if>
      </xacro:if>
    </xacro:if>

  </ros2_control>

  <!-- Load the gazebo_ros2_control plugin if using Gazebo hardware -->
  <xacro:if value="${hardware_type == 'gz_classic'}">
    <gazebo>
      <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
        <ros>
          <namespace>/$(arg robot_name)</namespace>
        </ros>
        <robot_param>robot_description</robot_param>
        <robot_param_node>robot_state_publisher</robot_param_node>
        <parameters>$(find interbotix_xsarm_sim)/config/trajectory_controllers/${robot_model}_trajectory_controllers.yaml</parameters>
      </plugin>
    </gazebo>
    <xacro:include filename="$(find interbotix_common_sim)/media/materials/scripts/interbotix_texture.gazebo"/>
  </xacro:if>

  <xacro:if value="${hardware_type == 'ignition'}">
    <gazebo>
      <plugin filename="libign_ros2_control-system.so" name="ign_ros2_control::IgnitionROS2ControlPlugin">
        <ros>
          <namespace>/$(arg robot_name)</namespace>
        </ros>
        <robot_param>robot_description</robot_param>
        <robot_param_node>robot_state_publisher</robot_param_node>
        <parameters>$(find arm_sim_scenario)/config/trajectory_controllers/${robot_model}_trajectory_controllers.yaml</parameters>
      </plugin>
    </gazebo>
    <xacro:include filename="$(find arm_sim_scenario)/meshes/interbotix_texture.gazebo"/>
  </xacro:if>

</robot>


